lib/datasource/go/goproxy.ts

```diff
@@ -0,0 +1,192 @@
+import { GetReleasesConfig, Release, ReleaseResult } from '../types';
+import { http } from './common';
+import { GoproxyFallback, GoproxyHost, VersionInfo } from './types';
+
+export function encodeCase(input: string): string {
+  return input.replace(/([A-Z])/g, (x) => `!${x.toLowerCase()}`);
+}
+
+export async function listVersions(
+  baseUrl: string,
+  lookupName: string
+): Promise<string[]> {
+  const url = `${baseUrl}/${encodeCase(lookupName)}/@v/list`;
+  const { body } = await http.get(url);
+  return body
+    .split(/\s+/)
+    .filter(Boolean)
+    .filter((x) => x.indexOf('+') === -1);
+}
+
+export async function versionInfo(
+  baseUrl: string,
+  lookupName: string,
+  version: string
+): Promise<Release> {
+  const url = `${baseUrl}/${encodeCase(lookupName)}/@v/${version}.info`;
+  const res = await http.getJson<VersionInfo>(url);
+
+  const result: Release = {
+    version: res.body.Version,
+  };
+
+  if (res.body.Time) {
+    result.releaseTimestamp = res.body.Time;
+  }
+
+  return result;
+}
+
+/**
+ * @example
+ * // [
+ * //   { url: 'foo', fallback: '|' },
+ * //   { url: 'bar', fallback: ',' },
+ * //   { url: 'baz', fallback: '|' },
+ * // ]
+ * parseGoproxy('foo|bar,baz')
+ */
+export function parseGoproxy(input: unknown): GoproxyHost[] | null {
+  if (!input || !is.string(input)) {
+    return null;
+  }
+
+  const result: GoproxyHost[] = input
+    .split(/([^,|]*(?:,|\|))/)
+    .filter(Boolean)
+    .map((s) => s.split(/(?=,|\|)/))
+    .map(([url, separator]) => ({
+      url,
+      fallback:
+        separator === ','
+          ? GoproxyFallback.WhenNotFoundOrGone
+          : GoproxyFallback.Always,
+    }));
+
+  return result.length ? result : null;
+}
+
+// https://golang.org/pkg/path/#Match
+const pathGlobLexer = {
+  main: {
+    separator: {
+      match: /\s*?,\s*?/,
+      value: (_: string) => '|',
+    },
+    asterisk: {
+      match: '*',
+      value: (_: string) => '[^\\/]*',
+    },
+    qmark: {
+      match: '?',
+      value: (_: string) => '[^\\/]',
+    },
+    characterRangeOpen: {
+      match: '[',
+      push: 'characterRange',
+      value: (_: string) => '[',
+    },
+    char: /[^*?\\[\n]/,
+    escapedChar: {
+      match: /\\./,
+      value: (s: string) => s.slice(1),
+    },
+  },
+  characterRange: {
+    char: /[^\\\]\n]/,
+    escapedChar: {
+      match: /\\./,
+      value: (s: string) => s.slice(1),
+    },
+    characterRangeEnd: {
+      match: ']',
+      pop: 1,
+    },
+  },
+};
+
+export function parseNoproxy(input: unknown): RegExp | null {
+  if (!input || !is.string(input)) {
+    return null;
+  }
+  const lexer = moo.states(pathGlobLexer);
+  lexer.reset(input);
+  const noproxyPattern = [...lexer].map(({ value }) => value).join('');
+  return noproxyPattern ? regEx(noproxyPattern) : null;
+}
+
+export function getProxyList(
+  proxy: string = process.env.GOPROXY,
+  noproxy: string = process.env.GONOPROXY || process.env.GOPRIVATE
```

Comment:
Are these documented in Go as being "or" ? i.e. you can't have both and combine them?

Reply:
`GOPRIVATE` acts as a default value for `GONOPROXY` and `GONOSUMDB`

Reply:
Renamed to `parseGoproxyEnv()`, as it accounts for both `GOPROXY` and disables results matching `GONOPRPOXY` pattern